### DISCUSSION TITLE
Why are my controller parameters horrible?

### DISCUSSION THREAD

**Discussion by Moradielham** [yalmip/YALMIP#1135] [Problems using YALMIP]
Dear Professor, I'm simulating an article that uses state feedback to control power electronic converters. I have attained the desired gamma value and the system in this state is stable, feasible and optimal but the control coefficients are strangely large (in the range of 1e8) which as you know is a very large value for duty cycle. I did not succeed by applying various constraints on the controller. Which part do you think the problem is? Why are the controller coefficients so large?

**Comment by johanlofberg** [comment]
H-infinity controllers for instance can have the property that the optimal controller sends some poles to -infinity which requires arbitrarily large feedback in the limit. To avoid this, you have to compute marginally sub-optimal controllers, by for instance penalizing various decision variables, or in the case of the standard L = inv(P)Y variable change limit the smallest eigenvalue of P and/or penalize it to force it to become significant, or most easily, solve the problem to optimality, and then solve a feasibility problem where the performance in limited to be within, say, 5% of optimality, by resolving with optimize([constraints, objective <= 1.05*value(objective)]. With most solver, you will then get a nicely behaved solution

**Reply by johanlofberg** [reply]
impossible to answer as we don't know what you actually do.

Of course there is a problem as you take the inverse and thus get huge values. For the optimality-relaxed problem, I doubt P would have eigenvalues close to zero, so I doubt that you did it correctly

**Reply by johanlofberg** [reply]
just post it here

**Comment by Moradielham** [comment]


**Reply by johanlofberg** [reply]
Huh?

**Reply by johanlofberg** [reply]
If you look above, all you see is a string in persian, no link or code

**Reply by johanlofberg** [reply]
What I told you to do is

```
% Solve to get optimal gamma
optimize(Cons,Gamma)
% resolve to find a solution which isn't necessarily optimal, but good, and hope solver returns nice solutions in interior
% instead of pushing things to feasibility border, when we simply look for feasible solution
optimize([Cons, Gamma <= 1.05*value(Gamma)])
```

Smallest eigenvalue goes from 10^-10 (i.e. true optimal solution is singular, causing K to be infinite). In the second run where solver will return a solution on the central path thus in the interior, smallest eigenvalue is 2e-4 and K will be in the order of thousands. 

Trying to explicitly push up the smallest eigenvalue further 

```
optimize(Cons,Gamma)
sdpvar t
optimize([Cons, X >= t*eye(length(X)), Gamma <= 1.05*value(Gamma)],-t)
```

yields a smallest eigenvalue of 4e-4. Forcing this eigenvalue could have led to detrimental effects on Y getting large at optimality and counteracting the smaller inverse, but that turned out to not happen, and K drops yet a bit in magnitude and is now in the hundreds instead.

**Reply by johanlofberg** [reply]
Why *should* it be in the range of 1e-3? Since minimizing the norm of the state-feedback is hard (most likely an intractable problem) you will have to use some heuristic. As you know it is constructed from inverse of X and Y, some kind of objective which combines -gamma (forcing X to be large thus inverse large) and norm(Y) of some kind would be a reasonable start.

**Reply by johanlofberg** [reply]
I have no idea what ku+0/7<1 means

**Reply by johanlofberg** [reply]
in most control notation, you would have u = Kx (input =controller*state) so i don't understand what you are talking about, nor do i see what +0/7 is supposed to be as it is 0 so why even write it, as all you have then is ku<1, and from your previous text I would then assume u is a constant vector containing number in the order of 10^3 or something. very unclear. you also talk about both K and k

**Reply by johanlofberg** [reply]
There is no magic choice of penalties and compromise parameters. You hack around until you find a strategy which works. In your case though, I suspect you are trying to achieve something impossible, either due to incorrect modelling, following an incorrect paper, or misunderstanding something.

Papers can be full of errors, typos and undocumented stuff to get things to work.

**Reply by johanlofberg** [reply]
Considering that the paper was published recently, I suspect some author still is alive and available.

**Reply by johanlofberg** [reply]
Why would you assume that? Would you not help if someone showed interest in your published research
